src/riak_cs_bw_limitter.erl

```diff
@@ -0,0 +1,197 @@
+                    Error
+            end;
+        {error, {_, Current, Threshold}} = Error2 ->
+            lager:warning("User ~p has exceeded bandwidth limit: usage, limit = ~p, ~p (bytes/sec)",
+                          [OwnerKey, Current, Threshold]),
+            Error2
+    end.
+
+new_user_state(Table, User, #state{table=Table} = _) ->
+
+    UserState = #user_state{user = User},
+    lager:debug("quota init: ~p => ~p", [User, UserState]),
+    %% Here's a race condition where if so many concurrent access
+    %% come, each access can yield a new fresh state and then
+    %% receives not access limitation.
+    catch ets:insert_new(Table, UserState),
+    UserState.
+
+test_bandwidth_state(DurationSec,
+                     #user_state{bandwidth = Bandwidth},
+                     #state{bandwidth_max = BandwidthMax}) ->
+    case Bandwidth / DurationSec < BandwidthMax  of
+        true -> ok;
+        false when BandwidthMax < 0 -> ok; %% No quota set
+        false -> {error, {bandwidth, Bandwidth, BandwidthMax}}
+    end.
+
+test_access_state(#user_state{access_count = AccessCount},
+                  #state{access_count_max = AccessCountMax}) ->
+    case AccessCount < AccessCountMax of
+        true -> ok;
+        false when AccessCountMax < 0 -> ok; %% No quota set
+        false -> {error, {access_count, AccessCount, AccessCountMax}}
+    end.
+
+-spec update(binary(), wm_log_data(), #state{}) -> ok | {error, term()}.
+update(User,
+       #wm_log_data{response_length = ResponseLength,
+                    headers = Headers} = _LogData,
+       State) ->
+    #state{table=Table} = State,
+    Bytes0 = case is_integer(ResponseLength) of
+                true -> ResponseLength;
+                _ -> 0
+            end,
+    RequestLength = mochiweb_headers:get_value("content-length", Headers),
+    Bytes = case catch list_to_integer(RequestLength) of
+                Len when is_integer(Len) -> Len + Bytes0;
+                _ -> Bytes0
+            end,
+    try
+        case ets:update_counter(Table, User, [{3, Bytes}, {4, 1}]) of
+            [_, _] -> ok;
+            Error0 ->
+                lager:warning("something wrong? ~p", [Error0]),
+                {error, Error0}
+        end
+    catch
+        error:badarg -> %% record not just found here
+            lager:debug("Cache of ~p (maybe not found)", [User]),
+            ok;
+        Type:Error ->
+            %% TODO: show out stacktrace heah
+            _ = lager:warning("something wrong? ~p", [Error]),
+            {error, {Type, Error}}
+    end.
+
+handle_info(refresh, #state{table=Table, duration_sec=Sec} = State) ->
+    ets:delete_all_objects(Table),
+    _ = maybe_schedule_refresh(Sec),
+    {noreply, State};
+handle_info(_, State) ->
+    {noreply, State}.
+
+update_state(Data, #state{table=Table} = _State) ->
+    ets:delete_all_objects(Table),
+    {DurationSec, BandwidthMax, AccessCountMax} = Data,
+    _ = maybe_schedule_refresh(DurationSec),
```

Comment:
If this function is called 1000 times per sec, many refresh events will fire! :fire: